crates/neon-runtime/src/napi/tsfn.rs

```diff
@@ -0,0 +1,202 @@
+//! Idiomatic Rust wrappers for N-API threadsafe functions
+
+use std::ffi::c_void;
+use std::mem::MaybeUninit;
+
+use crate::napi::bindings as napi;
+use crate::raw::{Local, Env};
+
+unsafe fn string(env: Env, s: impl AsRef<str>) -> Local {
+    let s = s.as_ref();
+    let mut result = MaybeUninit::uninit();
+
+    assert_eq!(
+        napi::create_string_utf8(
+            env,
+            s.as_bytes().as_ptr() as *const _,
+            s.len(),
+            result.as_mut_ptr(),
+        ),
+        napi::Status::Ok,
+    );
+
+    result.assume_init()
+}
+
+#[derive(Debug)]
+struct Tsfn(napi::ThreadsafeFunction);
+
+unsafe impl Send for Tsfn {}
+unsafe impl Sync for Tsfn {}
+
+#[derive(Debug)]
+/// Threadsafe Function encapsulate a Rust function pointer and N-API threadsafe
+/// function for scheduling tasks to execute on a JavaScript thread.
+pub struct ThreadsafeFunction<T> {
+    tsfn: Tsfn,
+    callback: fn(Env, T),
+}
+
+#[derive(Debug)]
+struct Callback<T> {
+    callback: fn(Env, T),
+    data: T,
+}
+
+/// Error returned when scheduling a threadsafe function with some data
+pub struct CallError<T> {
+    kind: napi::Status,
+    data: T,
+}
+
+impl<T> CallError<T> {
+    /// The specific error that occurred
+    pub fn kind(&self) -> napi::Status {
```

Comment:
Do we want to expose any `napi::` APIs? It might be better to have our own Neon enum for this. (We can change this in followup work, though, since this isn't yet stabilized.)

Reply:
It's not exposed in `neon` only in `neon-runtime`. I agree that we should have a separate enum, but it's not used until we add bounded queues.